### DISCUSSION TITLE
Is there any specification for .env?

### DISCUSSION THREAD

**Discussion by powerman** [dotenv-linter/dotenv-linter#805] [Q&A]
At a glance it looks like different tools/libraries supports different sets of features in `.env` (e.g. quotes, variable interpolation, `export` prefix, etc.) and some of them may conflict (meaning different tools may parse same `.env` with different results).

Is there any specification for `.env` which may ends this chaos?
Same question other way: is dotenv-linter supports only safe subset of `.env` features guaranteed to have same results no matter which tool/library will parse it after successful linting?

**Comment by mstruebing** [comment]
I think the closest to a specification you can get is this here, by the creator of the original dotenv nodejs package which is widely used: https://dotenvx.com/docs/env-file.

I think, in the end what is right or wrong all boils down to what the shell you are using would do with these env file.
I'm not sure there is a specification for it, probably just like an unwritten standard that's used, but as you are able to just source an env-file it should be dependent on the shell. 

Very vague, but I hope that helps.

**Reply by powerman** [reply]
`dotenvx` is great _(if we don't bother about 45MB file size and ¼sec execution time)_ but it looks like just one more (incompatible with others) implementation. I mean, if you always use `dotenvx` and never try to load `.env` using other tools/libraries it'll work great, but if you'll occasionally try to run, say, `docker compose` and it'll load `.env` then everything will breaks (because `docker compose` won't support neither command substitution nor encryption nor …). Worse, this breakage in some cases may go unnoticed.

Maybe I get things wrong, but in my understanding the initial idea behind `.env` is to get rid of custom shell scripts and wrappers (yeah, ones like `dotenvx`), and make it "just works" everywhere automatically out of box. And this succeeds - to some minimal extend where `A=B` and `A="B with spaces"` is enough to our needs. But for most projects it's just not enough… :disappointed: And here we add features not compatible with all other tools/libraries… which kills initial idea of `.env`.

Why do we need `dotenvx` as a 45MB slow wrapper to be installed everywhere if we already have such a wrapper: `/bin/sh`? `sh` starts much (about 250 times) faster and already here (and even if it's not - `dash` is just 100K-1MB depending on dynamic/static linking).

Don't get me wrong - I'm honestly trying to understand: what's the point in **that** `.env`? Without common spec more or less supported by all implementations it looks useless and even harmful to me. Why not use `env.sh` instead - this one actually "just works".

**Reply by mstruebing** [reply]
> Maybe I get things wrong, but in my understanding the initial idea behind .env is to get rid of custom shell scripts and wrappers (yeah, ones like dotenvx), and make it "just works" everywhere automatically out of box.

Can you give some source of this? I don't think that's the idea behind `.env`.
As far as I understand, `.env` is just a file that **by convention** defines some environment variables and can be sourced/used by your tooling/automatically or manually with direnv for example: https://direnv.net/#basic-installation.

Primary use is to define environment variables, but you could do all kind of stuff with it.
It's just a convention and not enforced by anything.

I may be totally wrong here, but that's what I understand and think a `.env`-file is used for.

**Reply by powerman** [reply]
> Can you give some source of this?

Looking for a sources leads me to an interesting archeological research. From what I was able to find it looks like first `.env` mention was in @ddollar tool foreman, introduced in https://github.com/ddollar/foreman/commit/3b8fec463d9966c5dc119c67084780849a219cd7. Then other tools like @bkeepers dotenv emerge and refer to foreman: https://github.com/bkeepers/dotenv/commit/c3568a06b341f1182bd4e8b0d6e58a594cac7966.

So, it looks like choosing cool but not unique configuration file name for one tool results in hijacking that cool name by a lot of other tools. And each of these tools believe it's the only tool who uses that file and thus may define it format and features at will. Sometimes this leads to inconsistencies even within one tool's ecosystem, e.g. https://github.com/docker/cli/issues/3630.

The problem is most (if not all) these tool loads this file by default. Which means user has no control over this - any other command executed by some command started by user may decide to parse this file in that tool's own unique way. They may have even different behaviour related to priority for values defined in `.env` file and in actual environment. So, even if we run `dotenvx run -- ./some-service` there is no guarantee that "some-service" won't re-read `.env` on start, parse it in different way, and use these (incorrectly parsed) values over real environment provided by `dotenvx` tool.

All of this sounds a bit scary to me - we can get unexpected behaviour and spend a lot of time debugging such an issue (like it already happened for many in mentioned above docker/cli issue).

But you're right - it looks like there was no real "idea" to let different tools share same `.env` file. It was just an usual chaos, nothing more.

**Comment by bkeepers** [comment]
:wave:

> Looking for a sources leads me to an interesting archeological research. From what I was able to find it looks like first `.env` mention was in @ddollar tool foreman, introduced in [ddollar/foreman@3b8fec4](https://github.com/ddollar/foreman/commit/3b8fec463d9966c5dc119c67084780849a219cd7). Then other tools like @bkeepers dotenv emerge and refer to foreman: [bkeepers/dotenv@c3568a0](https://github.com/bkeepers/dotenv/commit/c3568a06b341f1182bd4e8b0d6e58a594cac7966).

Thanks @powerman for pinging me. This part of the history is accurate: I created the first `dotenv` inspired by foreman.

My goal was to conform to `sh`/`bash` as much as possible with syntax, variable substitions, and command substitution. Beyond that, there is not a formal spec.

**Comment by ccoVeille** [comment]
It's good to see such a discussion in this repository. It gives historical context about how `.env` format emerged.

I have only one remark.

Here it's a repository about a dotenv linter.

The tool is great, and maintained. Both are important, and as valuable 🥰

For me, the maintainer is not accountable about the lack of spec about .env

Yes, it would have been better to have a clear specification. But there is apparently none.

So, yes, it's difficult in a linter to define what should be reported or not.

The default value, or the variable expansion for examples.

Yes, other tools like dotenvx supports a lot of things, and its size is a problem.

But here again, I'm a bit annoyed to feel like there are complains here about things the maintainer of the current tool where the discussion happens is not accountable, nor responsible, nor something he could do something about.

So here to me, I feel like there is kind of anger or at least frustration, shared in the posts, while once again. There is nothing we can do about the situation.

I might have misunderstood the posts, but it's how I felt them.

Share the passion, and the love.

Respectfully,

A random guy.

**Reply by powerman** [reply]
> There is nothing we can do about the situation.

Actually, there are some things we can do about it.

If it turns out there are no spec and each app parse `.env` in it's own way, and we're discussing this in a linting context, then linter may take this fact in account. For example, linter may need to require user to provide which tool he use to read `.env` in linter's args and then add 2 extra checks:
- validate `.env` does not use features not supported by any of listed tools
- validate there are no differences between listed tools in parsing `.env` (at least for subset of dotenv features used in current `.env`)

Another thing is more general but also may worth trying: we should write the spec and push all tools/libraries to support it. And probably add some kind of versioning header into `.env` to make it clear is it conform to the spec.

Or, if everything above will fail, we can consider using `.env` harmful and add a warning to notify users about this.
